File: src/main/java/net/softbell/bsh/component/PermissionComp.java
package net.softbell.bsh.component;

import java.util.ArrayList;
import java.util.List;

import org.springframework.stereotype.Component;

import lombok.RequiredArgsConstructor;
import net.softbell.bsh.domain.EnableStatusRule;
import net.softbell.bsh.domain.GroupRole;
import net.softbell.bsh.domain.entity.GroupPermission;
import net.softbell.bsh.domain.entity.Member;
import net.softbell.bsh.domain.entity.MemberGroup;
import net.softbell.bsh.domain.entity.MemberGroupItem;
import net.softbell.bsh.domain.entity.Node;
import net.softbell.bsh.domain.entity.NodeGroup;
import net.softbell.bsh.domain.entity.NodeGroupItem;
import net.softbell.bsh.domain.repository.GroupPermissionRepo;
import net.softbell.bsh.domain.repository.MemberGroupItemRepo;
import net.softbell.bsh.domain.repository.MemberGroupRepo;
import net.softbell.bsh.domain.repository.NodeGroupItemRepo;
import net.softbell.bsh.domain.repository.NodeGroupRepo;

/**
 * @Author : Bell(bell@softbell.net)
 * @Description : 권한 관련 컴포넌트
 */
@RequiredArgsConstructor
@Component
public class PermissionComp
{
	// Global Field
	private final GroupPermissionRepo groupPermissionRepo;
	private final MemberGroupRepo memberGroupRepo;
	private final MemberGroupItemRepo memberGroupItemRepo;
	private final NodeGroupRepo nodeGroupRepo;
	private final NodeGroupItemRepo nodeGroupItemRepo;
	
	// 활성화된 사용자 그룹 반환
	public List<MemberGroup> getEnableMemberGroup()
	{
		// Field
		List<MemberGroup> listMemberGroup;
		
		// Init
		listMemberGroup = memberGroupRepo.findByEnableStatus(EnableStatusRule.ENABLE);
		
		// Return
		return listMemberGroup;
	}
	
	// 사용자가 포함된 활성화된 사용자 그룹 반환
	public List<MemberGroup> getEnableMemberGroup(Member member)
	{
		// Field
		List<MemberGroupItem> listMemberGroupItem;
		List<MemberGroup> listMemberGroup;
		
		// Init
		listMemberGroupItem = memberGroupItemRepo.findByMember(member);
		listMemberGroup = memberGroupRepo.findByMemberGroupItemsInAndEnableStatus(listMemberGroupItem, EnableStatusRule.ENABLE);
		
		// Return
		return listMemberGroup;
	}
	
	// 권한이 있는 사용자 그룹 반환
	public List<MemberGroup> getPrivilegeMemberGroup(List<GroupPermission> listGroupPermission)
	{
		// Return
		return memberGroupRepo.findByGroupPermissionsIn(listGroupPermission);
	}
	
	// 권한이 있는 사용자 그룹 반환
	public List<MemberGroup> getPrivilegeMemberGroup(List<MemberGroup> listMemberGroup, List<GroupPermission> listGroupPermission)
	{
		// Field
		List<Long> listMemberGroupId;
		
		// Init
		listMemberGroupId = new ArrayList<Long>();
		
		// Load
		for (MemberGroup entity : listMemberGroup)
			listMemberGroupId.add(entity.getMemberGroupId());
		
		// Return
		return memberGroupRepo.findByMemberGroupIdInAndGroupPermissionsIn(listMemberGroupId, listGroupPermission);
	}
	
	// 활성화된 노드 그룹 반환
	public List<NodeGroup> getEnableNodeGroup()
	{
		// Field
		List<NodeGroup> listNodeGroup;
		
		// Init
		listNodeGroup = nodeGroupRepo.findByEnableStatus(EnableStatusRule.ENABLE);
		
		// Return
		return listNodeGroup;
	}
	
	// 노드가 포함된 활성화된 노드 그룹 반환
	public List<NodeGroup> getEnableNodeGroup(Node node)
	{
		// Field
		List<NodeGroupItem> listNodeGroupItem;
		List<NodeGroup> listNodeGroup;
		
		// Init
		listNodeGroupItem = nodeGroupItemRepo.findByNode(node);
		listNodeGroup = nodeGroupRepo.findByNodeGroupItemsInAndEnableStatus(listNodeGroupItem, EnableStatusRule.ENABLE);
		
		// Return
		return listNodeGroup;
	}
	
	// 권한이 있는 사용자 그룹 반환
	public List<NodeGroup> getPrivilegeNodeGroup(List<NodeGroup> listNodeGroup, List<GroupPermission> listGroupPermission)
	{
		// Field
		List<Long> listNodeGroupId;
		
		// Init
		listNodeGroupId = new ArrayList<Long>();
		
		// Load
		for (NodeGroup entity : listNodeGroup)
			listNodeGroupId.add(entity.getNodeGroupId());
		
		// Return
		return nodeGroupRepo.findByNodeGroupIdInAndGroupPermissionsIn(listNodeGroupId, listGroupPermission);
	}
	
	
	// 사용자 그룹으로 특정 권한이 포함된 그룹 권한 반환
	public List<GroupPermission> getMemberGroupPermission(GroupRole role, List<MemberGroup> listMemberGroup)
	{
		// Field
		List<GroupPermission> listGroupPermission;
		
		// Init
		listGroupPermission = groupPermissionRepo.findByGroupPermissionAndMemberGroupIn(role, listMemberGroup);
		
		// Return
		return listGroupPermission;
	}
	
	// 노드 그룹으로 특정 권한이 포함된 그룹 권한 반환
	public List<GroupPermission> getNodeGroupPermission(GroupRole role, List<NodeGroup> listNodeGroup)
	{
		// Field
		List<GroupPermission> listGroupPermission;
		
		// Init
		listGroupPermission = groupPermissionRepo.findByGroupPermissionAndNodeGroupIn(role, listNodeGroup);
		
		// Return
		return listGroupPermission;
	}
	
	// 노드 그룹과 사용자 그룹으로 특정 권한이 포함된 그룹 권한 반환
	public List<GroupPermission> getGroupPermission(GroupRole role, List<MemberGroup> listMemberGroup, List<NodeGroup> listNodeGroup)
	{
		// Field
		List<GroupPermission> listGroupPermission;
		
		// Init
		listGroupPermission = groupPermissionRepo.findByGroupPermissionAndMemberGroupInAndNodeGroupIn(role, listMemberGroup, listNodeGroup);
		
		// Return
		return listGroupPermission;
	}
}

File: src/main/kotlin/net/softbell/bsh/component/PermissionComp.kt
package net.softbell.bsh.component

import net.softbell.bsh.domain.EnableStatusRule
import net.softbell.bsh.domain.GroupRole
import net.softbell.bsh.domain.entity.*
import net.softbell.bsh.domain.repository.*
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component
import java.util.*

/**
 * @author : Bell(bell@softbell.net)
 * @description : 권한 관련 컴포넌트
 */
@Component
class PermissionComp {
    // Global Field
    @Autowired lateinit var groupPermissionRepo: GroupPermissionRepo
    @Autowired lateinit var memberGroupRepo: MemberGroupRepo
    @Autowired lateinit var memberGroupItemRepo: MemberGroupItemRepo
    @Autowired lateinit var nodeGroupRepo: NodeGroupRepo
    @Autowired lateinit var nodeGroupItemRepo: NodeGroupItemRepo

    // 활성화된 사용자 그룹 반환
    fun getEnableMemberGroup(): List<MemberGroup> {
        // Return
        return memberGroupRepo.findByEnableStatus(EnableStatusRule.ENABLE)
    }

    // 사용자가 포함된 활성화된 사용자 그룹 반환
    fun getEnableMemberGroup(member: Member): List<MemberGroup> {
        // Field
        val listMemberGroup: List<MemberGroup>

        // Init
        val listMemberGroupItem: List<MemberGroupItem> = memberGroupItemRepo.findByMember(member)
        listMemberGroup = memberGroupRepo.findByMemberGroupItemsInAndEnableStatus(listMemberGroupItem, EnableStatusRule.ENABLE)

        // Return
        return listMemberGroup
    }

    // 권한이 있는 사용자 그룹 반환
    fun getPrivilegeMemberGroup(listGroupPermission: List<GroupPermission>): List<MemberGroup> {
        // Return
        return memberGroupRepo.findByGroupPermissionsIn(listGroupPermission)
    }

    // 권한이 있는 사용자 그룹 반환
    fun getPrivilegeMemberGroup(listMemberGroup: List<MemberGroup>, listGroupPermission: List<GroupPermission>): List<MemberGroup> {
        // Field
        val listMemberGroupId: MutableList<Long>

        // Init
        listMemberGroupId = ArrayList()

        // Load
        for (entity in listMemberGroup) listMemberGroupId.add(entity.memberGroupId)

        // Return
        return memberGroupRepo.findByMemberGroupIdInAndGroupPermissionsIn(listMemberGroupId, listGroupPermission)
    }

    // 활성화된 노드 그룹 반환
    fun getEnableNodeGroup(): List<NodeGroup> {
        // Return
        return nodeGroupRepo.findByEnableStatus(EnableStatusRule.ENABLE)
    }

    // 노드가 포함된 활성화된 노드 그룹 반환
    fun getEnableNodeGroup(node: Node): List<NodeGroup> {
        // Field
        val listNodeGroup: List<NodeGroup>

        // Init
        val listNodeGroupItem: List<NodeGroupItem> = nodeGroupItemRepo.findByNode(node)
        listNodeGroup = nodeGroupRepo.findByNodeGroupItemsInAndEnableStatus(listNodeGroupItem, EnableStatusRule.ENABLE)

        // Return
        return listNodeGroup
    }

    // 권한이 있는 사용자 그룹 반환
    fun getPrivilegeNodeGroup(listNodeGroup: List<NodeGroup>, listGroupPermission: List<GroupPermission>): List<NodeGroup> {
        // Field
        val listNodeGroupId: MutableList<Long>

        // Init
        listNodeGroupId = ArrayList()

        // Load
        for (entity in listNodeGroup) listNodeGroupId.add(entity.nodeGroupId)

        // Return
        return nodeGroupRepo.findByNodeGroupIdInAndGroupPermissionsIn(listNodeGroupId, listGroupPermission)
    }


    // 사용자 그룹으로 특정 권한이 포함된 그룹 권한 반환
    fun getMemberGroupPermission(role: GroupRole, listMemberGroup: List<MemberGroup>): List<GroupPermission> {
        // Return
        return groupPermissionRepo.findByGroupPermissionAndMemberGroupIn(role, listMemberGroup)
    }

    // 노드 그룹으로 특정 권한이 포함된 그룹 권한 반환
    fun getNodeGroupPermission(role: GroupRole, listNodeGroup: List<NodeGroup>): List<GroupPermission> {
        // Return
        return groupPermissionRepo.findByGroupPermissionAndNodeGroupIn(role, listNodeGroup)
    }

    // 노드 그룹과 사용자 그룹으로 특정 권한이 포함된 그룹 권한 반환
    fun getGroupPermission(role: GroupRole, listMemberGroup: List<MemberGroup>, listNodeGroup: List<NodeGroup>): List<GroupPermission> {
        // Return
        return groupPermissionRepo.findByGroupPermissionAndMemberGroupInAndNodeGroupIn(role, listMemberGroup, listNodeGroup)
    }
}